### PR TITLE
projector: fix `RunPaths` result aliasing

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
The projector uses `multiplexer.RunPaths()` to find each run’s logdir on
disk, but it’s not clear from the code how this stays up to date as more
runs are loaded. The answer is insidious: `RunPaths` returns the raw
underlying dictionary, not a shallow copy of it, so changes to the
internal structures of the multiplexer are immediately reflected in the
projector plugin. This is a bit terrifying, since it means both that the
code is wildly non-threadsafe and that it’s hard to tell when the data
has actually changed. (The terror is somewhat abated by remembering that
concurrency safety around dicts in Python is all just pretend, anyway.)

This patch modifies the projector plugin to always shallow-copy the
result of `RunPaths`, and makes the necessary follow-up changes to
observe updates to the data. The result is that `RunPaths` is called
more times, up to once on each HTTP handler or `is_active` call. This is
fine: it’s expected that plugin routes may hit the multiplexer or data
provider, and since the projector plugin only really works locally, this
should be fast enough to not matter.

In doing so, we also improve the update semantics, since we can test
precisely for cache invalidation rather than just looking at the count
of runs. There are almost certainly still observable concurrency issues;
this patch doesn’t try to fix them all.

The eventual goal of this series of patches is to remove the dependency
on the multiplexer entirely, so the call to `RunPaths` will become a
call to a data provider’s `list_runs` and merging in `flags.logdir`.

Test Plan:
The projector plugin still works in the normal case (shows embeddings
with sprites, etc.). Edge cases to check:

  - Point TensorBoard at a logdir containing only non-projector data.
    Launch TensorBoard and wait a bit for the projector to finish
    determining that it has no data. Then add one projector run to the
    logdir and refresh TensorBoard (but do not restart the server). Note
    that the projector plugin properly picks up the new run. Repeat, and
    note that it picks up the newer run, too (i.e., the cache can be
    updated even when it’s non-empty).

  - Point TensorBoard at a logdir where the `projector_config.pbtxt` is
    inside a logdir’s plugin assets directory:

    ```
    logs/run/plugins/org_tensorflow_tensorboard_projector/projector_config.pbtxt
    ```

    Note that the config is still resolved properly. You can tell that
    it’s using the plugin assets because if you comment out the
    `_append_plugin_asset_directories` call, the embeddings will still
    be displayed but will lack sprite information.

wchargin-branch: projector-fix-aliasing

### CODE REVIEW

**Review by psybuzz — APPROVED** [review]
Manually checked the edge cases in the description and confirmed things work.

> Point TensorBoard at a logdir where the projector_config.pbtxt is
inside a logdir’s plugin assets directory: logs/run/plugins/org_tensorflow_tensorboard_projector/projector_config.pbtxt

I got hung up for a bit on this, because after I created that directory structure, TB wasn't detecting the config for some reason.  Adding an empty file named with the substring "tfevents" in that directory made things work, presumably because it forced the multiplexer to recognize that directory.

File: tensorboard/plugins/projector/projector_plugin.py [hunk 1]
```diff
@@ -325,44 +323,52 @@ def frontend_metadata(self):
...
 
+    def _update_run_paths(self):
+        """Updates `self._run_paths`, testing for updates.
+
+        Returns:
+          `True` if this call changed the value of `self._run_paths`,
+          else `False`.
+        """
+        if self.multiplexer:
+            run_paths = dict(self.multiplexer.RunPaths())
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/plugins/projector/projector_plugin.py:334]: Probably, yes, but we can only do that once we’ve ensured that it’s
[safe]. If we’d done so before this commit, the projector plugin would
have broken. So before making that change, I’d like to do at least a
cursory audit, and I’d like that to not be tied to this commit.

[safe]: https://www.hyrumslaw.com/

**Review by wchargin — COMMENTED** [review]
> > Point TensorBoard at a logdir where the projector_config.pbtxt is
> > inside a logdir’s plugin assets directory: logs/run/plugins/org_tensorflow_tensorboard_projector/projector_config.pbtxt
> 
> I got hung up for a bit on this, because after I created that directory structure, TB wasn't detecting the config for some reason. Adding an empty file named with the substring "tfevents" in that directory made things work, presumably because it forced the multiplexer to recognize that directory.

Hmm—you only need an event file at the run directory, right?

```
$ find /tmp/proj/mnist/lr*
/tmp/proj/mnist1/lr_1E-03,conv=1,fc=2
/tmp/proj/mnist1/lr_1E-03,conv=1,fc=2/events.out.tfevents.1563406327.HOSTNAME
/tmp/proj/mnist1/lr_1E-03,conv=1,fc=2/plugins
/tmp/proj/mnist1/lr_1E-03,conv=1,fc=2/plugins/org_tensorflow_tensorboard_projector
/tmp/proj/mnist1/lr_1E-03,conv=1,fc=2/plugins/org_tensorflow_tensorboard_projector/projector_config.pbtxt
$ bazel run //tensorboard -- --logdir /tmp/proj  # works
```

You specifically should not need it for the plugin assets directory
itself: i.e., `projector_config.pbtxt` does not need to have any event
files as siblings.

**Comment by wchargin** [comment]
Also, if you *do* need an event file in the plugin assets directory for
some reason (I can’t reproduce this), could you please check whether you
observe the same behavior against head?

**Comment by psybuzz** [comment]
> You specifically should not need it for the plugin assets directory
itself: i.e., projector_config.pbtxt does not need to have any event
files as siblings.

Just double checked and can confirm it's as you say.  In a scenario with
- run dir having a tfevents file
- plugin assets dir has only projector_config.pbtxt (no siblings)
the Projector still properly picks up the config file.

**Review by psybuzz — APPROVED** [review]

**Comment by wchargin** [comment]
Okay; great. Thanks.
